src/Coordination/KeeperStateMachine.h

```diff
@@ -182,8 +156,105 @@ class KeeperStateMachine : public nuraft::state_machine
 
     KeeperSnapshotManagerS3 * snapshot_manager_s3;
 
-    KeeperStorage::ResponseForSession processReconfiguration(
-        const KeeperStorage::RequestForSession& request_for_session)
-        TSA_REQUIRES(storage_and_responses_lock);
+    virtual KeeperStorageBase::ResponseForSession processReconfiguration(
+        const KeeperStorageBase::RequestForSession& request_for_session)
+        TSA_REQUIRES(storage_and_responses_lock) = 0;
+
+};
+
+/// ClickHouse Keeper state machine. Wrapper for KeeperStorage.
+/// Responsible for entries commit, snapshots creation and so on.
+template<typename Storage>
```

Comment:
Emmmmm... I will try it

Reply:
Wow, I remember I have tried my best to make template as few as possible. But
- almost all class contains `KeeperStorage`
- we cannot implement `IKeeperStorage` because there are lots of fields like `sessions_and_watchers` / `ephemerals` being called directly.